parity-clib/parity.hpp

```diff
@@ -0,0 +1,228 @@
+#error "The Parity Ethereum C++ bindings require C++17 or newer.  Please upgrade your compiler, or specify the needed flags to enable C++17 support."
+#endif
+#include <cassert>
+#include <exception>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <parity.h>
+#include <stdexcept>
+#include <string>
+#include <vector>
+#include <type_traits>
+static_assert(std::is_same<uintptr_t, size_t>::value, "Rust requires that uintptr_t and size_t are the same type");
+
+namespace parity {
+// avoid conflict with other Parity projects
+namespace ethereum {
+using namespace std::literals::string_literals;
+
+class ParityException final : public std::exception {};
+
+class ParityLogger final {
+  friend class ParityEthereum;
+  parity_logger *logger;
+
+public:
+  ParityLogger(const std::string &log_mode, const std::string &log_file)
+      : logger(nullptr) {
+    parity_set_logger(log_mode.size() ? log_mode.data() : nullptr,
+                      log_mode.size(),
+                      log_file.size() ? log_file.data() : nullptr,
+                      log_file.size(), &this->logger);
+  }
+  ParityLogger(const ParityLogger &other) = delete;
+  ParityLogger &operator=(const ParityLogger &other) = delete;
+  ParityLogger(ParityLogger &&other) noexcept : logger(nullptr) {
+    this->logger = other.logger;
+    other.logger = nullptr;
+  }
+  ParityLogger &operator=(ParityLogger &&other) noexcept {
+    if (this != &other) {
+      this->logger = other.logger;
+      other.logger = nullptr;
+    }
+    return *this;
+  }
+  ~ParityLogger() {
+    assert(nullptr == this->logger &&
+           "ParityLogger objects must be moved into a ParityParams, "
+           "not destroyed");
+  }
+};
+
+class ParityConfig final {
+  friend class ParityEthereum;
+  parity_config *config;
+
+public:
+  ParityConfig(const ParityConfig &other) = delete;
+  ParityConfig &operator=(const ParityConfig &other) = delete;
+  ParityConfig(const std::vector<std::string> &cli_args) : config(nullptr) {
+    size_t const size = cli_args.size();
+    std::vector<size_t> len_vecs;
+    std::vector<char const *> args;
+    len_vecs.reserve(size);
+    args.reserve(size);
+    for (const auto &i : cli_args) {
+      len_vecs.push_back(i.size());
+      args.push_back(i.data());
+    }
+    if (parity_config_from_cli(size ? args.data() : nullptr,
+                               size ? len_vecs.data() : nullptr, size, &config))
+      throw std::runtime_error(
+          "failed to create Parity Ethereum configuration");
+  }
+  ParityConfig(ParityConfig &&other) noexcept : config(nullptr) {
+    this->config = other.config;
+    other.config = nullptr;
+  }
+  ParityConfig &operator=(ParityConfig &&other) noexcept {
+    if (this != &other) {
+      this->config = other.config;
+      other.config = nullptr;
+    }
+    return *this;
+  }
+  ~ParityConfig() {
+    if (this->config)
+      assert(false && "ParityConfig objects must be moved into a "
```

Comment:
This will trigger if an exception is thrown before the user moves the `ParityCOnfig` into a `ParityParams`.
What's wrong with just calling `parity_config_destroy` here?

Reply:
Nothing, but there is no corresponding `parity_logger_destroy`.

Reply:
Fixed